tajo-core/src/main/java/org/apache/tajo/worker/LocalFetcher.java

```diff
@@ -0,0 +1,451 @@
+    this.host = uri.getHost() == null ? "localhost" : uri.getHost();
+    this.port = uri.getPort();
+    if (port == -1) {
+      if (scheme.equalsIgnoreCase("http")) {
+        this.port = 80;
+      } else if (scheme.equalsIgnoreCase("https")) {
+        this.port = 443;
+      }
+    }
+
+    bootstrap = new Bootstrap()
+        .group(
+            NettyUtils.getSharedEventLoopGroup(NettyUtils.GROUP.FETCHER,
+                conf.getIntVar(ConfVars.SHUFFLE_RPC_CLIENT_WORKER_THREAD_NUM)))
+        .channel(NioSocketChannel.class)
+        .option(ChannelOption.ALLOCATOR, NettyUtils.ALLOCATOR)
+        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS,
+            conf.getIntVar(ConfVars.SHUFFLE_FETCHER_CONNECT_TIMEOUT) * 1000)
+        .option(ChannelOption.SO_RCVBUF, 1048576) // set 1M
+        .option(ChannelOption.TCP_NODELAY, true);
+  }
+
+  public LocalFetcher(TajoConf conf, URI uri, ExecutionBlockContext executionBlockContext, String tableName) {
+    super(conf, uri);
+    this.localFileSystem = executionBlockContext.getLocalFS();
+    this.localDirAllocator = executionBlockContext.getLocalDirAllocator();
+    this.maxUrlLength = conf.getIntVar(ConfVars.PULLSERVER_FETCH_URL_MAX_LENGTH);
+    this.tableName = tableName;
+
+    Optional<TajoPullServerService> optional = executionBlockContext.getSharedResource().getPullServerService();
+    if (optional.isPresent()) {
+      // local pull server service
+      this.pullServerService = optional.get();
+      this.host = null;
+      this.bootstrap = null;
+
+    } else if (PullServerUtil.useExternalPullServerService(conf)) {
+      // external pull server service
+      pullServerService = null;
+
+      String scheme = uri.getScheme() == null ? "http" : uri.getScheme();
+      this.host = uri.getHost() == null ? "localhost" : uri.getHost();
+      this.port = uri.getPort();
+      if (port == -1) {
+        if (scheme.equalsIgnoreCase("http")) {
+          this.port = 80;
+        } else if (scheme.equalsIgnoreCase("https")) {
+          this.port = 443;
+        }
+      }
+
+      bootstrap = new Bootstrap()
+          .group(
+              NettyUtils.getSharedEventLoopGroup(NettyUtils.GROUP.FETCHER,
+                  conf.getIntVar(ConfVars.SHUFFLE_RPC_CLIENT_WORKER_THREAD_NUM)))
+          .channel(NioSocketChannel.class)
+          .option(ChannelOption.ALLOCATOR, NettyUtils.ALLOCATOR)
+          .option(ChannelOption.CONNECT_TIMEOUT_MILLIS,
+              conf.getIntVar(ConfVars.SHUFFLE_FETCHER_CONNECT_TIMEOUT) * 1000)
+          .option(ChannelOption.SO_RCVBUF, 1048576) // set 1M
+          .option(ChannelOption.TCP_NODELAY, true);
+    } else {
+      endFetch(FetcherState.FETCH_FAILED);
+      throw new TajoInternalError("Pull server service is not initialized");
+    }
+  }
+
+  @Override
+  public List<FileChunk> get() throws IOException {
+    return pullServerService != null ? getDirect() : getFromFetchURI();
```

Comment:
There are two cases for local fetcher. 
* When an internal pull server is running, local fetchers can retrieve data directly. 
* When an external pull server is running, 
 * If the shuffle type is hash, local fetchers can still retrieve data directly.
 * If the shuffle type is range, local fetchers need to get meta information of data via HTTP. Once the meta information is retrieved, they can read data directly. 

I'll add this description.